lib/src/main/java/io/stormbird/token/tools/TokenScriptValidator.java

```diff
@@ -0,0 +1,219 @@
+    //
+    //    where "document" is the name of a file containing the XML document
+    //    to be validated.
+    //
+    public static SignatureCheck check(InputStream fileStream) throws Exception
+    {
+        SignatureCheck result = new SignatureCheck();
+
+        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
+        dbFactory.setNamespaceAware(true);
+        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
+        Document xml = dBuilder.parse(fileStream);
+        xml.getDocumentElement().normalize();
+
+        // Find Signature element
+        NodeList nl =
+                xml.getElementsByTagNameNS(XMLSignature.XMLNS, "Signature");
+        if (nl.getLength() == 0)
+        {
+            //throw new Exception("Cannot find Signature element");
+            return result;
+        }
+
+        // Create a DOM XMLSignatureFactory that will be used to unmarshal the
+        // document containing the XMLSignature
+        XMLSignatureFactory fac = XMLSignatureFactory.getInstance("DOM");
+
+        // Create a DOMValidateContext and specify a KeyValue KeySelector
+        // and document context
+        DOMValidateContext valContext = new DOMValidateContext
+                (new CertKeySelector(), nl.item(0));
+
+        // unmarshal the XMLSignature
+        XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+
+        // Validate the XMLSignature (generated above)
+        result.isValid = signature.validate(valContext);
+
+        if (result.isValid)
+        {
+            KeyInfo kInfo = signature.getKeyInfo();
+            X509CertImpl cert = selectKeyData(kInfo.getContent());
+            result.issuerPrincipal = cert.getIssuerX500Principal().getName();
+            result.subjectPrincipal = cert.getSubjectX500Principal().getName();
+            result.keyType = cert.getSigAlgName();
+
+            for (Object o : kInfo.getContent())
+            {
+                XMLStructure xmlStructure = (XMLStructure) o;
+                if (xmlStructure instanceof KeyName)
+                {
+                    result.keyName = ((KeyName) xmlStructure).getName();
+                }
+            }
+        }
+        else
+        {
+            System.err.println("Signature failed core validation");
+            boolean sv = signature.getSignatureValue().validate(valContext);
+            System.out.println("signature validation status: " + sv);
+            // check the validation status of each Reference
+            Iterator i = signature.getSignedInfo().getReferences().iterator();
+            for (int j = 0; i.hasNext(); j++)
+            {
+                boolean refValid =
+                        ((Reference) i.next()).validate(valContext);
+                System.out.println("ref[" + j + "] validity status: " + refValid);
+            }
+        }
+
+        return result;
+    }
+
+    protected static X509CertImpl selectKeyData(List list)
```

Comment:
@JamesSmartCell these names are very confusing for us.

Reply:
> @JamesSmartCell these names are very confusing for us.

Before you blame anyone, think that not everybody gets your advantage of being born in an English speaking country.

Oh wait a minute, you are referring to James?

Reply:
Hmm? I clearly marked @JamesSmartCell not sure what you are referring to.